src/toxcore/state_format/old.rs

```diff
@@ -597,68 +614,52 @@ pub const FRIENDSTATEBYTES: usize = 1      // "Status"
                                   + 8;     // last time seen
 
 impl FromBytes<FriendState> for FriendState {
-    fn from_bytes(bytes: &[u8]) -> Option<Self> {
-        if bytes.len() < FRIENDSTATEBYTES { return None }
+    fn parse_bytes(bytes: &[u8]) -> ParseResult<Self> {
+        if bytes.len() < FRIENDSTATEBYTES {
+            return parse_error!("Not enough bytes for FriendState.")
+        }
 
-        let status = match FriendStatus::from_bytes(&bytes) {
-            Some(s) => s,
-            None => return None,
-        };
+        let Parsed(status, bytes) = try!(FriendStatus::parse_bytes(bytes));
 
-        const PK_POS: usize = 1;
         let pk = match PublicKey::from_slice(
-                            &bytes[PK_POS..PUBLICKEYBYTES + PK_POS]) {
+                            &bytes[..PUBLICKEYBYTES]) {
             Some(pk) => pk,
-            None => return None,
+            None => return parse_error!("Can't parse PublicKey.")
         };
+        let bytes = &bytes[PUBLICKEYBYTES..];
 
-        // get string out of bytes in range (start, start+len]
-        // supply start position and *length*
-        let get_string = |start, len: usize| -> Option<String> {
+        // parse string out of bytes
+        // supply length
+        fn parse_string(bytes: &[u8], len: usize) -> ParseResult<String> {
             let str_len = u16::from_be(array_to_u16(
-                        &[bytes[start+len], bytes[start+len+1]])) as usize;
-            String::from_utf8(bytes[start..start+str_len].to_vec()).ok()
+                        &[bytes[len], bytes[len+1]])) as usize;
+            match String::from_utf8(bytes[..str_len].to_vec()).ok() {
```

Comment:
perhaps `ok()` could be removed? Without it, `Err(e)` could provide more useful info for `parse_error!`